Sprint 2

el-mariachi opened the pull request:
Мы строили строили, и наконец построили!

el-mariachi committed:
style: Add Eslint to project. Basic setup

el-mariachi committed:
style: Add Stylelint to project

el-mariachi committed:
chore: Edit scripts, eslintignore, edit tsconfig

Change tsconfig.json: remove typeroots,
add ../typings/*.d.ts to include paths. Remove package.json from
prettierignore. Separate out tsconfig.netlify.json for deploy.

el-mariachi committed:
feat: Add EventBus & Block components

el-mariachi committed:
feat: Inroduce testing entry for parcel

el-mariachi committed:
test: Create a test button module

el-mariachi committed:
feat: Update Block class to emit CDU only once

Even if more than one prop gets changed.
Change filename block.ts to Block.ts.

el-mariachi committed:
feat: Add/remove event handlers in Block

Update Block class to add/remove event handlers for this._element
in _render() method.

el-mariachi committed:
feat: Add UUID to Block data-id attribute

el-mariachi committed:
fix: Update setProps() & CDU in Block

When updating props with setProps() store the result in this._meta.
Otherwise it would be old-old props. Change _componentDidUpdate()
to emit render event instead of calling render() directly.
Add ownKeys() to props proxy to omit _ props from iteration.

el-mariachi committed:
feat: Render child components inside parent Block

Introduce hierarchical component structure in Block.ts. Separate props
from children in Block constructor. Insert stubs into component template
and replace stubs with child elements later (insertChildren() method).
Update _render() to return element instead of markup.
Update _componentDidMount to dispach CDM to children.
Update render() default implementation.
Update typings.d.ts to use imported handlebars type.

el-mariachi committed:
feat: Update children when setting props in Block

Update setProps() method to separate props from children the same way
it is done in the constructor. Set this.children to accomodate
incoming children with Object.assign().

el-mariachi committed:
refactor: Update EventBus implementation

Take a nicer version of EventBus from
https://dev.to/openhacking/how-to-implement-an-event-bus-in-javascript-15io
Modify to suit our needs. Create a separate Singleton version.

el-mariachi committed:
feat: Add HTTPTransport class

el-mariachi committed:
chore: Improve project folder structure

el-mariachi committed:
feat: Complete HTTPTransport class

el-mariachi committed:
feat: Make page 404 use Block

el-mariachi committed:
feat: Remove wrapper elmnt from Block, use on 404

Each Block must define a meaningful root element and it's
classes and attributes in the constructor method. The root element of
that block must be removed from the template and transferred to code.

el-mariachi committed:
feat: Update 500 page to use no-wrapper Block

el-mariachi committed:
feat: Make Block accept arrays of children

Methods updated: _componentDidMount, _getChildren, setProps,
insertChildren. We can now use Handlebars's 'each' helper on children.

el-mariachi committed:
feat: Make login form use children array w/inputs

Decompose Form into InputGroup an Input components.

el-mariachi committed:
feat: Make Button component, fix input focus

el-mariachi committed:
refactor: Simplify Block setProps & proxy children

el-mariachi committed:
feat: Reveal form inputs by name

el-mariachi committed:
refactor: Use Array.reduce to create inputs object

Form.ts method init(). Get rid of Map().

el-mariachi committed:
WIP on sprint_2: 217da2e refactor: Use Array.reduce to create inputs object

el-mariachi committed:
index on sprint_2: 217da2e refactor: Use Array.reduce to create inputs object

el-mariachi committed:
undo: Revert Form back to using Map

el-mariachi committed:
feat: Add validation to form inputs, make 2 forms

ProfileForm and ComposeForm.

el-mariachi committed:
feat: Complete login form with validation

el-mariachi committed:
feat: Add sign-up page with components

el-mariachi committed:
feat: Prepare for events[] prop

el-mariachi committed:
feat: Make Block & Form accept array for an event

el-mariachi committed:
feat: Merge events in InputGroup.

el-mariachi committed:
feat: Validate repeat password in Form

el-mariachi committed:
fix: Add static link to edit_ava, change formName

el-mariachi committed:
refactor: Change insertChildren to compile

el-mariachi committed:
feat: Add InputError component use global eventbus

el-mariachi committed:
feat: Decouple Input and Form

el-mariachi committed:
Finish form validation & login page

el-mariachi committed:
feat: Finish sign_up page with validation

el-mariachi committed:
chore: Set aside edit_profile static

el-mariachi committed:
feat: Start edit_profile page new

el-mariachi committed:
feat: Finish edit_profile, with file input static

el-mariachi committed:
chore: Shash edit_ava edit_profile_static ->svalka

el-mariachi committed:
refactor: Add "Page" to page classe's names

el-mariachi committed:
feat: Create Search component

el-mariachi committed:
feat: Add ChatListHeader & Link components

el-mariachi committed:
feat: Add Menu component

el-mariachi committed:
feat: Add Chat component w/select, chat-list page

el-mariachi committed:
refactor: Rename ChatArea -> MessageArea globally

el-mariachi committed:
chore: Clean up style imports

el-mariachi committed:
feat: Add MessageAreaHeader component

el-mariachi committed:
feat: Add Messages, link to chat_list, break vars

el-mariachi committed:
feat: Finish chat page, fix form names

el-mariachi committed:
chore: Cleanup pages

el-mariachi committed:
docs: Edit README.md

el-mariachi committed:
fix: Place MessageAreaHeader behind #if

el-mariachi committed:
fix: Edit netlify.toml

el-mariachi committed:
refactor: Revert to relative paths

el-mariachi committed:
fix: Get rid of import.meta.url

el-mariachi committed:
fix: Get rid of ~/ globally

el-mariachi committed:
refactor: Replace new URL() with image import

el-mariachi committed:
style: Fix some styling errors

el-mariachi committed:
refactor: Set back to old EventBus+singleton

el-mariachi committed:
refactor: Bring back advanced EventBus

Remove optional _meta.props in Block, make chat & chat_list pages
inherit from Page, which loads Page.scss.

el-mariachi committed:
chore: Cleanup index

el-mariachi committed:
refactor: Move chat list children to page level

el-mariachi committed:
fix: Avoid excessive code splitting in build

el-mariachi committed:
docs: Edit README.md

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)


К сожалению, работу придется отклонить без потери итерации:

* не могу запустить проект локально. Если `npm run dev`, то ничего не отображается на `http://localhost:1234` , а `npm start` выдает ошибку  https://disk.yandex.ru/i/gmqAsfgKnLXC5g   или `npm run start:local` тоже ничего не дает. Вы `server.ts` не доделали, чтобы он раздавал данные


Еще заметил ошибку:

*  Нужно обернуть список с ссылками в тег `nav` по семантике, так как это навигация по сайту

el-mariachi committed:
review: Add nav to index, combine server code

el-mariachi commented:
Добрый день! 

Спасибо, что оперативно смогли посмотреть мою работу!

Мой `server.ts` работает и раздает данные. В первой строке делается импорт `import { app } from './src/server/app'`;
Но, чтобы у вас не было вопросов, я теперь слил в один файл.
Почему` npm start` дает ошибку, я понимаю: нужно сначала запустить `npm run build`, о чем написано в README.md. Жаль, что это никто не читает, хотя от нас требуют этот документ оформлять! Но мой `npm start:local`, который вы увидели в `package.json,` у меня работает и без предварительной сборки.
Скрипт `npm start` запускает `node dist/server` — скомпилированный тайпскриптом js-файл. Это для Netlify.
Скрипт` npm start:local` запускает исходный ts-файл командой `ts-node server.ts` в корне проекта. Какие тут могут быть сложности?
Также в README.md есть ссылка на деплой в Netlify, который часто представляет собой отдельный челендж (у меня, к счастью, трудностей не было почти), и который у меня работает. 
Я попросил коллег по курсу проверить мою сборку. Откликнулись два человека, одни с windows, другой с Linux. У обоих `npm run dev` собирает и запускает проект.
Пожалуйста, проверьте еще раз, может, у вас что-то поехало? ))
Отправляю повторно на проверку с косметическими правками (см. 7290f86).
Написал в README.md алгоритм запуска проекта (c2a481a).

el-mariachi committed:
docs: Edit README.md

el-mariachi committed:
docs: Edit README.md, add node version

el-mariachi committed:
refactor: Factor out mock data

Make Chats component and MessageArea component be responsible for
rendering their contents independently via event bus.

gennady-bars commented:
> Почему` npm start` дает ошибку, я понимаю: нужно сначала запустить `npm run build`, о чем написано в README.md.

по идее, нужно объединять несколько скриптов в один, чтобы при `npm start` запускались команды очистки папки, `build` и запуск сервера. Не стоит вручную прогонять все скрипты по отдельности

gennady-bars commented on package.json:
```diff
@@ -28,16 +35,31 @@
     "ts-node": "10.9.1",
-    "typescript": "4.8.4"
+    "typescript": "4.8.4",
+    "uuid": "9.0.0"
```
Это тоже нужно в `dependencies` перенести

В `devDependencies` указывается то, что необходимо для сборки и разработки приложения, а в `dependencies` нужно указывать то, что необходимо непосредственно для работы приложения (express, uuid). Все линтеры, тесты и тд нужно только в `devDependencies` указывать, чтобы зря не раздувать кол-во пакетов в собранном для деплоя приложении.

gennady-bars commented on package.json:
```diff
@@ -28,16 +35,31 @@
     "@parcel/transformer-inline-string": "2.7.0",
     "@parcel/transformer-sass": "2.7.0",
     "@types/express": "4.17.14",
+    "@types/uuid": "^8.3.4",
```
`    "@types/uuid": "^8.3.4",
`

Лучше задавать **фиксированные версии** зависимостям. Если удалить `package-lock.json` и попытаться повторно установить зависимости с помощью команды `npm install`, то **npm** установит последние версии зависимостей, что может привести к ошибкам.

gennady-bars commented on static/styles/vars.scss:
```diff
@@ -18,3 +18,8 @@ $modalBack: rgba(0, 0, 0, 0.6);
 // shadows
```
* Отлично, что цвета вынесены в `css`-константы

gennady-bars commented on src/components/Message/Message.hbs:
```diff
@@ -0,0 +1,10 @@
+<div class="Message-Bubble">
+  {{#if image}}
+  <div class="Message-Body Message_type_image">
+    <img class="Message-Image" src="{{ image.src }}" alt="{{ image.alt }}">
```
* Отлично, что не забываете про `alt` в тегах `img`

gennady-bars commented on typings/png.d.ts:
```diff
@@ -0,0 +1,4 @@
+declare module '*.png' {
+  const value: any;
```
типизация у изображений обычно `string`

```js

  const value: string;


```

gennady-bars commented on src/utils/HTTPTransport.ts:
```diff
@@ -0,0 +1,107 @@
+    .join('&');
+}
+export class HTTPTransport {
+  get(url: string, options: OptionsWithoutMethod = {}): Promise<XMLHttpRequest> {
```
**Можно лучше**
Если типизация повторяется в каждом методе, то нужно типизировать сам метод, а не дублировать типизацию аргументов. Код станет чище

```jsx

// создаем тип метода
type HTTPMethod = (url: string, options?: OptionsWithoutMethod) => Promise<XMLHttpRequest>

export class HTTPTransport {

  // используем тип и удаляем дублирование в аргументах
  get: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.GET}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  put: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.PUT}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  post: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.POST}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  delete: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.DELETE}, options.timeout)
  )

```

el-mariachi replied:
Из документации typescript:

[Arrow Functions](https://www.typescriptlang.org/docs/handbook/2/classes.html#arrow-functions)

- This will use more memory, because each class instance will have its own copy of each function defined this way
- You can’t use `super.<method-name>` in a derived class, because there’s no entry in the prototype chain to fetch the base class method from

А использование стандартных методов класса (а не свойств в виде стрелочных функций) требует явного задания типов параметров и возвращаемого значения, насколько я понял из своих изысканий. Хотелось бы использовать интерфейс, но не помогает.

gennady-bars commented on src/utils/HTTPTransport.ts:
```diff
@@ -0,0 +1,107 @@
+    const { method, headers, data } = parseOptions(options);
+
+    if (method === METHOD.GET && data) {
+      url = `${url}?${queryStringify(data)}`;
```
```js
      url = `${url}?${queryStringify(data)}`;


```


это можно было прямо в `get` делать на 45й строчке, чтобы не проверять при любых запросах

gennady-bars commented on src/classes/Block.ts:
```diff
@@ -0,0 +1,330 @@
+    hasID?: boolean;
+  };
+}
+export abstract class Block {
```
лучше всего передавать тип пропсов в виде дженерика в `Block`, чтобы можно было конкретный тип пропсов всегда иметь под каждый отдельный экземпляр класса `Block`. Будет очень универсально 

```js
class abstract Block<Props extends Record<string, any> = any> {

```

gennady-bars commented on src/utils/addInputHandlers.ts:
```diff
@@ -0,0 +1,23 @@
+import inputBlurHandler from '../controllers/inputBlurHandler';
+import inputFocusHandler from '../controllers/inputFocusHandler';
+
+const addedHandlers: EventsProp = {
```
Отлично, что упаковали сразу 2 обработчика в объект и не дублируете нигде больше

gennady-bars commented on src/pages/edit_profile/edit_profile.ts:
```diff
@@ -1 +1,176 @@
+  inputs,
+  buttons,
+  events: {
+    submit: submitForm,
```
Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку. Обработка события `submit` еще срабатывает при нажатии на `Enter`, так как это встроенная логика сабмитов в браузере.

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.


* `Readme` отлично оформлен
* Проект задеплоен и отлично работает
* Отлично, что не забыли про `.gitignore`
* Отлично, что ограничиваете версию `node`
* Отлично, что фиксируете версии зависимостей (почти все)
* Хорошая структура папок и файлов
* Добавлен `ESLint` и `Stylelint`
* Отлично, что цвета вынесены в `css`-константы
* Отлично, что не забываете про `alt` в тегах `img`
* Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку.

Удачного дальнейшего обучения.

* теперь все ок с `npm run dev`. Все запустилось
* В `devDependencies` указывается то, что необходимо для сборки и разработки приложения, а в `dependencies` нужно указывать то, что необходимо непосредственно для работы приложения (express, uuid). Все линтеры, тесты и тд нужно только в `devDependencies` указывать, чтобы зря не раздувать кол-во пакетов в собранном для деплоя приложении.
* Лучше задавать **фиксированные версии** зависимостям. Если удалить `package-lock.json` и попытаться повторно установить зависимости с помощью команды `npm install`, то **npm** установит последние версии зависимостей, что может привести к ошибкам.

el-mariachi committed:
review: Incorporate reviewer's comments sprint_2

el-mariachi committed:
Merge branch 'main' into sprint_2
